packages/apollo-gateway/src/buildQueryPlan.ts

```diff
@@ -545,17 +550,86 @@ function completeField(
 
     parentGroup.otherDependentGroups.push(...subGroup.dependentGroups);
 
+    let definition: FragmentDefinitionNode;
+    let selectionSet: SelectionSetNode;
+
+    if (context.compressDownstreamRequests && subGroup.fields.length > 2) {
+      ({ definition, selectionSet } = getInternalFragment(
+        subGroup.fields,
+        returnType,
+        context,
+      ));
+      parentGroup.internalFragments.add(definition);
+    } else {
+      selectionSet = selectionSetFromFieldSet(subGroup.fields, returnType);
+    }
+
+    // "Hoist" internalFragments of the subGroup into the parentGroup so all
+    // fragments can be included in the final request for the root FetchGroup
+    subGroup.internalFragments.forEach(fragment => {
+      parentGroup.internalFragments.add(fragment);
+    });
+
     return {
       scope,
       fieldNode: {
         ...fieldNode,
-        selectionSet: selectionSetFromFieldSet(subGroup.fields, returnType),
+        selectionSet,
       },
       fieldDef,
     };
   }
 }
 
+function getInternalFragment(
+  fields: Field<GraphQLCompositeType>[],
+  returnType: GraphQLCompositeType,
+  context: QueryPlanningContext
+) {
+  const expandedSelectionSet = selectionSetFromFieldSet(fields, returnType);
+
+  const key = print(expandedSelectionSet);
+  if (!context.internalFragments[key]) {
+    const name = `__QueryPlanFragment_${context.internalFragmentCount++}__`;
+
+    const definition: FragmentDefinitionNode = {
+      kind: 'FragmentDefinition',
+      name: {
+        kind: 'Name',
+        value: name,
+      },
+      typeCondition: {
+        kind: 'NamedType',
+        name: {
+          kind: 'Name',
+          value: returnType.name,
+        },
+      },
+      selectionSet: expandedSelectionSet,
+    };
+
+    const fragmentSelection: SelectionSetNode = {
+      kind: 'SelectionSet',
+      selections: [
+        {
+          kind: 'FragmentSpread',
+          name: {
+            kind: 'Name',
+            value: name,
+          },
+        },
+      ],
+    };
+    context.internalFragments[key] = {
+      name,
+      definition,
+      selectionSet: fragmentSelection,
+    };
+  }
+
```

Comment:
Maybe we could consider retaining the expanded selection set and updating a refcount here and then only emitting the fragment & definition if `count > some n`?

(Or, rather than an arbitrary n, we could actually check if `print(definition).length + count - count * print(expandedSelectionSet).length < 0` and only do the fragment if that's the case. Does that get messy with fragments including fragments? It seems like at worst it'll overestimate the savings.)

Reply:
I think since we're emitting these fragment spreads on the fly, we're pretty limited currently on when we get to emit.

I believe the approach you're hinting at requires a second pass one way or another - one which performs the analysis and another which handles some transformation. Certainly not opposed to something like this in a future refinement, but I'm not sure it fits into this approach.

Please let me know if I've misunderstood anything!